WIP Command Standard Streams

cindykimxp opened the pull request:
Please fill in the fields below to submit a pull request. The more information
that is provided, the better.

**Fixes issue #:**
in-toto#351, the issues presented in runlib.in-toto-run

**Description of the changes being introduced by the pull request:**
- To correctly print streams to console or record streams in link metadata at appropriate times
- Add quiet keyword argument (default True) to enable (False) printing streams to console
- Use record_streams keyword (default False) to enable (True) recording streams in link metadata

Follows the logic:
When record_streams is true:
- if quiet is true --> link metadata
- if quiet is false --> both link metadata and terminal

When record_streams is false:
- if quiet is true --> nowhere
- if quiet is false --> terminal

**Note:** changes to in-toto-run have not yet been made; there may be a need to add an argument under runlib.in_toto_run(...quiet=args.quiet)...
**Additional review requests:** "process.stdout" and "process.stderr"... a bit unsure if these are implemented correctly (info on these: https://nodejs.org/api/process.html#process_process_stderr) (https://nodejs.org/api/process.html#process_process_stdout)

Please verify and check that the pull request fulfills the following
requirements:
 - The code follows the Code Style Guidelines
 - Tests have been added for the bug fix or new feature
 - Docs have been added for the bug fix or new feature

Commit:
Prepare Debian 0.4.1-1 release

- Add 0.4.1-1 changelog entry
- Update securesystemslib dependency to >= 0.12.0

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

dependabot-preview[bot] committed:
Bump pathspec from 0.6.0 to 0.7.0

Bumps [pathspec](https://github.com/cpburnz/python-path-specification) from 0.6.0 to 0.7.0.
- [Release notes](https://github.com/cpburnz/python-path-specification/releases)
- [Changelog](https://github.com/cpburnz/python-path-specification/blob/master/CHANGES.rst)
- [Commits](https://github.com/cpburnz/python-path-specification/commits)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

SantiagoTorres committed:
roadmap-reviews: december '19

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: Santiago Torres <santiago@archlinux.org>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Fix securesystemslib min-dep in requirements.txt

Current in-toto requires securesystemslib>=0.12.0, which has the
gpg subpackage and process module that are no longer part of
in-toto proper.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
__init__.py: bump version to 0.4.2

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
CHANGELOG: add 0.4.2 entry

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Add Python 3.8 to tested versions

- Enable Py 3.8 for tox
- Enable Py 3.8 for travis
- Bump linting tox build Py version from 3.7 to 3.8
- Re-compile requirements-pinned.txt to combine pinned deps for
  all supported Python versions:
  - Removes obsolete transitive dep 'asn1crypto'
  - Updates securesystemslib to latest version
  - Re-adds transitive dependency subprocess
- Add Py 3.8 language classifier to setup.py

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Add custom sphinx argparse-example-epilog extension

The in_toto command line tools use argparse epilog to add an
example usage section to the help message. The epilog does not
render nicely when using the sphinx-argparse extension.

This commit adds a custom extension to properly render the title
(expected to be 'EXAMPLE USAGE') and code blocks of a command line
tools epilog.

See ArgparseUsageEpilog docstring for details about the extension.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Add argparse action group helper functions

Default argparse action groups are "optional arguments" and
"positional arguments". Most in-toto command line tools also use a
custom "required named arguments" action group.

This commit adds two helper function, one to make all group titles
title case, e.g. for consistency when using the sphinx argparse
extension, which title-cases default action groups only; and one
function to customize the order in which action groups are
displayed in the help message or in sphinx, to e.g. show required
arguments before optional arguments.

Note that the helpers access protected members of argparse.
Although not part of the official API these seem pretty stable, and
we don't heavily rely on them, in case they disappear, unit tests
will catch it and we can remove/replace the feature, for which they
are currently used.

This commit also adds tests for the two helpers.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Move argparser creation to separate function

Separate argument parser creation and configuration into a
create_parser function in all in-toto command line tools. This is
required to auto-generate sphinx documentation.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Customize cli action group titles and order

Call helper functions to make argparse action group titles tile
case and to create a action group order for all in-toto cli tools.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Enhance CLI descriptions, help and usage examples

Major overhaul of all CLI help texts, in an attempt to make them
more informative and render nicely and uniformly on
sphinx/readthedocs.

Kudos to @SantiagoTorres and his work in
https://github.com/in-toto/in-toto/pull/332, which I've been
integrating here.

Changes includes:

- add more usage examples to arpgarse.epilog field of cli tools
- Unify title and indentation in all usage examples to render nicely
  with the custom argparse-example-epilog sphinx extension
- uniformly start option help texts with lowercase
- uniformly start options with arguments with a noun (the argument)
  and options without arguments with a verb (what the option does)
- unify argument groups, i.e. use defaults (positional arguments,
  optional arguments) and one custom (required named arguments)
  where applicable
- uniformly use single quotes instead of backticks for emphasis in
  help texts
- move more common arguments (must be exactly the same!) to
  common_args.py

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Add sphinx auto cli doc configuration

Use sphinx-argparse and custom argparse-epilog sphinx extensions
to auto-generate cli documentation.

The directives provided by the two extensions are used in the newly
created doc/source/command-line-tools/in-toto-*.rst files, and
insert the output of the corresponding cli tool's help messages when
generating documentation with sphinx.

This commit further
- updates the table of context,
- adds a docs specific requirements file to install the 3rd-party
  extension)
- patches the path to make available the custom extension

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Add requirements file to build docs

Sphinx installs in-toto with its runtime dependencies, and
extra packages, e.g. the sphinx-argparse.

This file should be chosen via the readthedocs dashboard.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Manually check mutual exclusion in in-toto-record

Switch to manual mutual exclusion check for --key and --gpg
arguments (as in in_toto_run).

The `add_mutually_exclusive_group` used for the --key and --gpg
options can not be called on a custom argument group, such as the
"required named arguments" group. As a consequence the options
are listed in the "optional arguments" section. Since at least one
of them is required it is better to list both of them in the
"required named arguments" group (in addition to clarifying their
use in via the per-argument description.

This commit removes the use of add_mutually_exclusive_group and
siwtches to manual mutual exclusion check for --key and --gpg
arguments (as in in_toto_run) for better help message/cli docs.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Use kwargs in in-toto-run to fix lstrip bug

The in_toto_run cli does not support all of in_toto_run's
arguments. By passing the kwargs as args and omitting some
intermediate args the lstrip_paths argument is wrongly passed
as compact_json argument.

This commit fixes the bug by passing all kwargs with their keyword
from the in_toto_run cli to the corresponding function in runlib.
It also fixes the existing test.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Disambiguate gpg home dir and keyring

Some functions in securesystemslib.gpg support a homedir argument.
This corresponds to the gpg `--homedir` option, which is different
from the gpg `--keyring` option. However, in-toto documentation
inaccurately refers to the `homedir` as keyring in some places.
This commit fixes the relevant parts in the code and cli docs.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Remove help text from cli doc headers

Remove output of calling cli tools with the --help flag from all
cli modules' document headers as it needs to be copied/updated
manually and thus is prone to get out of sync.

The same (single-sourced) information can be found
- in the code,
- by installing the in-toto and running the cli command with
  the --help flag, and soon
- on in-toto.readthedocs.io

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Update cli documentation links in README.md

Point to readthedocs pages instead of document header of the
individual CLI modules.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Check type and not instance in test_common_args

When testing the undocumented (protected) attribute
'_action_groups' from 'argparse', it is safer to do a type and not
an instance check, because in 'common_args.sort_action_groups' we
replace what '_action_groups' references with a list. So if argparse
were to change its type to something list-ish, it could lead to
unexpected behavior on our side and we might not detect it in the
test here.

Kudos to @SantiagoTorres for pointing this out.

Pylint claims that this is an unidiomatic-typecheck. We beg to
differ.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

dependabot-preview[bot] committed:
Bump six from 1.13.0 to 1.14.0

Bumps [six](https://github.com/benjaminp/six) from 1.13.0 to 1.14.0.
- [Release notes](https://github.com/benjaminp/six/releases)
- [Changelog](https://github.com/benjaminp/six/blob/master/CHANGES)
- [Commits](https://github.com/benjaminp/six/compare/1.13.0...1.14.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Change order of argument sections in CLI docs

Change the section order of argument types to show "position
arguments" (required) before "optional arguments" (not required).

Further update in_toto_record to make the ordering and naming
consistent, following advice from @jhdalek55.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Refine install and CLI docs

Reinfe CLI and installation docs as shown on readthedocs following
an out-of-band review by @jhdalek55.

This mostly adds missing preposition and fixes wording.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Add TmpDirMixin for test cases

- Add TmpDirMixin with class methods to create and change to
  temporary directory, and to change back to original cwd and
  remove the temporary directory.

- Configure relevant test cases to use the mixin in their
  set up and tear down methods.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Add GPGKeysMixin for test cases

- Add GPGKeysMixin with classmethod to locate the test gpg rsa
  keyrings and copy them to the cwd (e.g. a temporary directory
  when used after TmpDirMixin.set_up_test_dir).

- Configure relevant test cases to use the mixin in their
  set up methods.

Note the mixin has a class variable gnupg_home which contains the
path to the keyring relative to cwd. This fixes an issue with gpg
on windows that has troubles migrating old keyrings if the homedir
is an absolute path with drive prefix (e.g. C:/path/to/keyring).

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Define gpg keyids for tests in GPGKeyMixin

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Remove unnused imports in tests

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

dependabot-preview[bot] committed:
Bump securesystemslib[crypto,pynacl] from 0.13.1 to 0.14.0

Bumps [securesystemslib[crypto,pynacl]](https://github.com/secure-systems-lab/securesystemslib) from 0.13.1 to 0.14.0.
- [Release notes](https://github.com/secure-systems-lab/securesystemslib/releases)
- [Changelog](https://github.com/secure-systems-lab/securesystemslib/blob/master/CHANGELOG.md)
- [Commits](https://github.com/secure-systems-lab/securesystemslib/compare/v0.13.1...v0.14.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

dependabot-preview[bot] committed:
Bump cffi from 1.13.2 to 1.14.0

Bumps [cffi](https://bitbucket.org/cffi/release-doc) from 1.13.2 to 1.14.0.
- [Commits](https://bitbucket.org/cffi/release-doc/commits)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

dependabot-preview[bot] committed:
Bump enum34 from 1.1.6 to 1.1.9

Bumps [enum34](https://bitbucket.org/stoneleaf/enum34) from 1.1.6 to 1.1.9.
- [Commits](https://bitbucket.org/stoneleaf/enum34/commits)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Polish installation instructions

Following @SantiagoTorres' review, this commit updates the
installation instructions to:
- disambiguate GPG (the tool) and PGP (the standard), and
- shorten another phrase.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

joshuagl committed:
Use constant from securesystemslib to detect GPG in tests

Use the new constant in securesystemslib 0.14.1 to check for the
presence of GPG while running tests.

Signed-off-by: Joshua Lock <jlock@vmware.com>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

dependabot-preview[bot] committed:
Bump pycparser from 2.19 to 2.20

Bumps [pycparser](https://github.com/eliben/pycparser) from 2.19 to 2.20.
- [Release notes](https://github.com/eliben/pycparser/releases)
- [Changelog](https://github.com/eliben/pycparser/blob/master/CHANGES)
- [Commits](https://github.com/eliben/pycparser/compare/release_v2.19...release_v2.20)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

dependabot-preview[bot] committed:
Bump enum34 from 1.1.9 to 1.1.10

Bumps [enum34](https://bitbucket.org/stoneleaf/enum34) from 1.1.9 to 1.1.10.
- [Commits](https://bitbucket.org/stoneleaf/enum34/commits)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

cindykimxp committed:
Command Standard Streams

Per default print standard streams to console but don't record them in link metadata
Use --quiet to disable printing streams to console
Use --record-streams to enable recording streams in link metadata

Signed-off-by: cindykimxp <cindykim100@gmail.com>

dependabot-preview[bot] committed:
Bump cryptography from 2.8 to 2.9

Bumps [cryptography](https://github.com/pyca/cryptography) from 2.8 to 2.9.
- [Release notes](https://github.com/pyca/cryptography/releases)
- [Changelog](https://github.com/pyca/cryptography/blob/master/CHANGELOG.rst)
- [Commits](https://github.com/pyca/cryptography/compare/2.8...2.9)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

adityasaky commented on in_toto/runlib.py:
```diff
@@ -319,7 +319,7 @@ def record_artifacts_as_dict(artifacts, exclude_patterns=None,
   return artifacts_dict
 
-def execute_link(link_cmd_args, record_streams):
+def execute_link(link_cmd_args, record_streams, quiet):
```
Is there a default for `quiet`?

adityasaky replied:
```suggestion
def execute_link(link_cmd_args, record_streams, quiet=True):
```

going by the call below

cindykimxp replied:
Oh, yes! I should've included this! Thanks for catching that

adityasaky commented on in_toto/runlib.py:
```diff
@@ -350,15 +350,29 @@ def execute_link(link_cmd_args, record_streams):
   if record_streams:
-    return_code, stdout_str, stderr_str = \
-        securesystemslib.process.run_duplicate_streams(link_cmd_args)
+    if (quiet == False): #record_streams true, quiet false
```
```suggestion
    if quiet is False: #record_streams true, quiet false
```
or
`if not quiet`
Alternatively, `if quiet:` with the body that you currently have under `else` should also work.

cindykimxp replied:
Yeah, I can definitely clean it up a bit. I'll move it around without the parentheses and the == (old school convention habits!) Thanks!

adityasaky commented on in_toto/runlib.py:
```diff
@@ -350,15 +350,29 @@ def execute_link(link_cmd_args, record_streams):
 
   else:
-    process = securesystemslib.process.run(link_cmd_args, check=False,
+    if (quiet == False): #record_streams false, quiet false
```
Same as above

adityasaky commented on in_toto/runlib.py:
```diff
@@ -350,15 +350,29 @@ def execute_link(link_cmd_args, record_streams):
+      securesystemslib.process.run_duplicate_streams(link_cmd_args)
+    else: #record_streams true, quiet true
+     process = securesystemslib.process.run(link_cmd_args, check=False,
+     stdout=securesystemslib.process.PIPE,
```
```suggestion
       stdout=securesystemslib.process.PIPE,
```

cindykimxp replied:
Oh, YES! I was looking for the correct way to do this one. I wasn't quite sure about this -- thank you!

adityasaky commented on in_toto/runlib.py:
```diff
@@ -350,15 +350,29 @@ def execute_link(link_cmd_args, record_streams):
+    else: #record_streams true, quiet true
+     process = securesystemslib.process.run(link_cmd_args, check=False,
+     stdout=securesystemslib.process.PIPE,
+     stderr=securesystemslib.process.PIPE)
```
```suggestion
       stderr=securesystemslib.process.PIPE)
```

adityasaky commented on in_toto/runlib.py:
```diff
@@ -350,15 +350,29 @@ def execute_link(link_cmd_args, record_streams):
+      return_code = process.returncode
+    else: #record_streams false, quiet true
+      process = securesystemslib.process.run(link_cmd_args, check=False,
       stdout=securesystemslib.process.DEVNULL,
```
```suggestion
        stdout=securesystemslib.process.DEVNULL,
```

adityasaky commented on in_toto/runlib.py:
```diff
@@ -350,15 +350,29 @@ def execute_link(link_cmd_args, record_streams):
+    else: #record_streams false, quiet true
+      process = securesystemslib.process.run(link_cmd_args, check=False,
       stdout=securesystemslib.process.DEVNULL,
       stderr=securesystemslib.process.DEVNULL)
```
```suggestion
        stderr=securesystemslib.process.DEVNULL)
```

adityasaky commented on in_toto/runlib.py:
```diff
@@ -528,7 +542,7 @@ def in_toto_run(name, material_list, product_list, link_cmd_args,
   if link_cmd_args:
     LOG.info("Running command '{}'...".format(" ".join(link_cmd_args)))
-    byproducts = execute_link(link_cmd_args, record_streams)
+    byproducts = execute_link(link_cmd_args, record_streams, quiet=True)
```
```suggestion
    byproducts = execute_link(link_cmd_args, record_streams, quiet=quiet)
```

adityasaky replied:
By using `quiet=True`, you're not actually using the `quiet` that was passed to `in_toto_run`, but rather setting a keyword arg for `execute_link`. :)

adityasaky reviewed (changes requested):
Thanks for the PR, @cindykimxp. Just some quick changes, which I think will also help with the CI errors.

adityasaky commented:
~Also, [`securesystemslib.process.run`](https://github.com/secure-systems-lab/securesystemslib/blob/master/securesystemslib/process.py#L55) calls [`subprocess.run`](https://docs.python.org/3/library/subprocess.html#subprocess.run). I suspect you'd have to pass `capture_output=True` with `stdout` and `stderr` to capture those streams. @lukpueh?~

Ah I think I misunderstood the condition initially, nvm.

adityasaky commented on in_toto/runlib.py:
```diff
@@ -350,15 +350,29 @@ def execute_link(link_cmd_args, record_streams):
+      return_code, stdout_str, stderr_str = \
+      securesystemslib.process.run_duplicate_streams(link_cmd_args)
+    else: #record_streams true, quiet true
+     process = securesystemslib.process.run(link_cmd_args, check=False,
```
Also, these have been indented with a single space instead of two.

cindykimxp replied:
I think I made the changes you suggested! Let me know if it hasn't correctly been updated -- and thanks so much!

cindykimxp committed:
WIP Command Standard Streams Update Suggestions

- Cleaned up the code with indentations 
- Default for quiet in execute_link
- Fixed passing in the quiet keyword into execute_link for byproducts instead of true

Signed-off-by: cindykimxp <cindykim100@gmail.com>

lukpueh reviewed (commented):
Many thanks for working on this, @cindykimxp! And thanks for your review, @adityasaky!

This already looks good. Here is a bit of feedback from my side:

- We need to make sure that the tests pass.
The reason why they only fail on Python 3 is that a Python 3 subprocess per default returns stdout and stderr with the wrong data type for our purposes.
I suggest you check out the `securesystemslib.process.run(...)` function and the documentation of the underlying Python function (hint: Use  node.js documentation won't help you much there).

- We should follow our code style guide
Please use double indentation (4 spaces) for line-continuation, this makes it easier to distinguish it from single indentation (2 spaces) that indicate a new block of code.

- (bonus) These are some high-level style considerations
You have four distinct blocks of code now in `execute_link`, that are executed depending on the combination of the values of `record_streams` and `quiet`. The latter three look fairly similar. Maybe you can do common things in those blocks only once? ... just like cancellation in math. :)

Thanks again for adding features to in-toto and please let me know if you have any questions.

dependabot-preview[bot] committed:
Bump pathspec from 0.7.0 to 0.8.0

Bumps [pathspec](https://github.com/cpburnz/python-path-specification) from 0.7.0 to 0.8.0.
- [Release notes](https://github.com/cpburnz/python-path-specification/releases)
- [Changelog](https://github.com/cpburnz/python-path-specification/blob/master/CHANGES.rst)
- [Commits](https://github.com/cpburnz/python-path-specification/commits)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

cindykimxp committed:
Update runlib.py

- Used 4 indents for line-continuation to follow code style guidelines
- Unindented the last two repeating lines of the else statement for cleanliness and simplicity

Signed-off-by: cindykimxp <cindykim100@gmail.com>

dependabot-preview[bot] committed:
Bump cryptography from 2.9 to 2.9.2

Bumps [cryptography](https://github.com/pyca/cryptography) from 2.9 to 2.9.2.
- [Release notes](https://github.com/pyca/cryptography/releases)
- [Changelog](https://github.com/pyca/cryptography/blob/master/CHANGELOG.rst)
- [Commits](https://github.com/pyca/cryptography/compare/2.9...2.9.2)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

dependabot-preview[bot] committed:
Bump six from 1.14.0 to 1.15.0

Bumps [six](https://github.com/benjaminp/six) from 1.14.0 to 1.15.0.
- [Release notes](https://github.com/benjaminp/six/releases)
- [Changelog](https://github.com/benjaminp/six/blob/master/CHANGES)
- [Commits](https://github.com/benjaminp/six/compare/1.14.0...1.15.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

SolidifiedRay committed:
Add a flag to accept a target directory to put generated metadata in

Added an extra optional flag '--metadata-directory' to the 'in-toto run' and the
'in-toto record stop' arguments; updated related methods in 'in_toto_record.py',
'in_toto_run.py', and 'runlib.py'; and added one test to 'test_in_toto_record.py',
one test to 'test_in_toto_run.py', and nine tests to 'test_runlib.py". If this flag
is passed, the resulting link metadata file will be dumped to the target directory.
Otherwise, the default destination is the current working directory.

Signed-off-by: SolidifiedRay <yc3346@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

dependabot-preview[bot] committed:
Bump pynacl from 1.3.0 to 1.4.0

Bumps [pynacl](https://github.com/pyca/pynacl) from 1.3.0 to 1.4.0.
- [Release notes](https://github.com/pyca/pynacl/releases)
- [Changelog](https://github.com/pyca/pynacl/blob/master/CHANGELOG.rst)
- [Commits](https://github.com/pyca/pynacl/compare/1.3.0...1.4.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

dependabot-preview[bot] committed:
Bump securesystemslib[crypto,pynacl] from 0.14.2 to 0.15.0

Bumps [securesystemslib[crypto,pynacl]](https://github.com/secure-systems-lab/securesystemslib) from 0.14.2 to 0.15.0.
- [Release notes](https://github.com/secure-systems-lab/securesystemslib/releases)
- [Changelog](https://github.com/secure-systems-lab/securesystemslib/blob/master/CHANGELOG.md)
- [Commits](https://github.com/secure-systems-lab/securesystemslib/compare/v0.14.2...v0.15.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

IshaVDave committed:
Add windows dev mode hint to symlink tests

Added an extra method '_raise_win_dev_mode_error' to the 'Test_ApplyExcl
udePatterns' class in test_runlib.py, and updated 3 other methods to inc
lude it . It checks if the user is on a Windows platform. If so, they ar
e informed that Developer mode is required to work with symlinks on Wind
ows, and asked if it is enabled.

Signed-off-by: Isha Dave <ishavdave@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Revise + consolidate docstring format and contents

Revise/shorten/consolidate/update main API function-, class- and
method- docstrings:

 - verifylib.in_toto_verify
 - runlib.in_toto_run
 - runlib.in_toto_record_start
 - runlib.in_toto_record_stop
 - models.metadata.Metablock
 - models.link.Link
 - models.layout.Layout
 - models.layout.Step
 - models.layout.Inspection
 - models.common.Signable (only parts relevant for read the docs)
 - models.common.ValidationMixin (only relevant parts)

Changes custom Secure Systems Lab docstring format to Google
format, which is machine- (sphinx) and human-readable.
(Choosing Google over NumPy because it requires less vertical
space, see
https://sphinxcontrib-napoleon.readthedocs.io/en/latest/#google-vs-numpy)

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Move settings docs from README.md to sphinx docs

+ revise and update

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Move layout creation sample code to sphinx docs

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Fix minor syntax issues in installation docs

Fix issues that occurred after swithing to sphinx-rtd-theme.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Update misc sphinx doc config

- Switch from alabaster to sphinx_rtd_theme
- Add napoleon sphinx extension to parse Google-style docstrings
- Add todo sphinx extension (for utils docs)
- Comment out currently unused html_static_path that triggered a
  warning

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Enable API + metadata model class sphinx auto-docs

+ some minor updates on the docs landing page, like pointers to
other in-toto resources.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Document metadata_directory arg + errors in runlib

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
Add final roadmap '20 review

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Co-authored-by: Aditya Sirish <8928778+adityasaky@users.noreply.github.com>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
__init__.py: bump version to 0.5.0

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

Commit:
CHANGELOG: add 0.5.0 entry

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>
Signed-off-by: cindykimxp <cindykim100@gmail.com>

cindykimxp committed:
Adding universal_newlines parameter

Python 3 tests were failing because of an error that did could not encode bytes.
To pass the tests for Python 3 (and not just Python 2), added in the universal_newlines parameter to get strings instead of bytes. 

Signed-off-by: Cindy Kim
Signed-off-by: cindykimxp <cindykim100@gmail.com>

cindykimxp committed:
Update runlib.py

Signed-off-by: cindykimxp <cindykim100@gmail.com>

cindykimxp committed:
Revert "Update runlib.py"

This reverts commit 2e043a6284d8435c7ce611c39c060174613e710e.

Signed-off-by: cindykimxp <cindykim100@gmail.com>

cindykimxp committed:
Merge branch 'develop' into cindykimxp-patch-1-1

Signed-off-by: cindykimxp <cindykim100@gmail.com>

cindykimxp committed:
Merge branch 'develop' into cindykimxp-patch-1-1

lukpueh commented on in_toto/runlib.py:
```diff
@@ -356,13 +356,25 @@ def execute_link(link_cmd_args, record_streams):
-        securesystemslib.process.run_duplicate_streams(link_cmd_args)
+    if not quiet: #record_streams true, quiet false
+      return_code, stdout_str, stderr_str = \
+      securesystemslib.process.run_duplicate_streams(link_cmd_args)
```
Please use double indentation here.

lukpueh commented on in_toto/runlib.py:
```diff
@@ -319,7 +319,7 @@ def record_artifacts_as_dict(artifacts, exclude_patterns=None,
   return artifacts_dict
 
-def execute_link(link_cmd_args, record_streams):
+def execute_link(link_cmd_args, record_streams, quiet=True):
```
I think `quiet` could be mandatory here. `execute_link` is more or less internal API and only called from `in_toto_run`. For `in_toto_run` OTOH `quiet` should indeed be optional.

lukpueh reviewed (commented):
💯  for figuring out the encoding issue! :) Here are some suggestions for next steps:

- Double indent line continuation (see inline comment below)
- Fix the linter issue flagged by Travis
- Add a description of the argument to the docstrings of `execute_link` and `in_toto_run`.
  NOTE: `execute_link` uses the old secure-systems-lab docstring style and `in_toto_run` uses the new google docstring style, please make it consistent with the existing styles.

cindykimxp committed:
Update runlib.py

- made quiet a mandatory parameter when running execute_link
- continued in_toto_run parameters in new line to avoid lint errors

cindykimxp committed:
Merge branch 'cindykimxp-patch-1-1' of https://github.com/cindykimxp/in-toto into cindykimxp-patch-1-1

Signed-off-by: cindykimxp <cindykim100@gmail.com>

cindykimxp committed:
Merge branch 'cindykimxp-patch-1-1' of https://github.com/cindykimxp/in-toto into cindykimxp-patch-1-1

Signed-off-by: cindykimxp <cindykim100@gmail.com>

cindykimxp committed:
Merge branch 'cindykimxp-patch-1-1' of https://github.com/cindykimxp/in-toto into cindykimxp-patch-1-1

cindykimxp committed:
Double Indentation and Clean Up 

- double indented a continuing line
- changes from quiet=quiet to just quiet

cindykimxp committed:
Merge branch 'cindykimxp-patch-1-1' of https://github.com/cindykimxp/in-toto into cindykimxp-patch-1-1

SantiagoTorres commented:
Closing this one in favor of #393
